Columns: Improve data handling/output and metadata export

flyth opened the pull request:
# Columns: Improve gadget data handling

This PR introduces a new library that can improve the way we handle our event/stats data in a couple of ways, described below.

More information about the functionality / features can be found here:
https://github.com/kinvolk/inspektor-gadget/blob/michael/tableformatter/pkg/columns/README.md

### Store metadata in the event/stats structs
Instead of having two or more places to define metadata like column widths, sorting parameters and so on, we can keep them right by the structs using tags. The library uses a combination of generics and reflection to achieve this. That way we can (in most cases) handle things like including/excluding columns, sorting, grouping automatically without writing additional code specific to the structs.

Special cases might need a bit of additional code, though.

### Optimize output for columns
The library is able to take into account column widths and other information from the tag metadata and will format the output accordingly. It supports
* shortening entries by either cutting at the width or indicating with an ellipsis ("...")
* adjust column widths automatically using the terminal width
* aligning values left or right

### Add helper functions
#### Sorting
Using the library one can simply sort the structs by one or more individual fields.

#### Grouping
Grouping by column content is also possible, so if you want for example group all entries from a specific node and return the sum of the values for another column (e.g. bpf programs' total runtime in top/ebpf on each node), you can simply do so without writing additional code.

#### Filtering
Different filter rules can be applied to sets of data, including: exact match, regular expression match, <, >, <=, >=.

### Metadata Export
All information gathered by the library from the structs can be exposed and used for other use cases. In case of command line handling via cobra it can return a list of all available columns without writing additional code. This could also be used for other frontends like web interfaces.

## Benchmarks
Depending on the use case, using reflection is usually slower than specific code. I did some basic benchmarks that show a x2-x5 decrease of performance for things like formatting a complete line, but as this is still happening in a couple of nanoseconds per row, it should be negligible for all our use cases.

## Todos
- [x] Extract example code for gadgets
- [x] Add unit tests

## Related Issues
#856, #851

mauriciovasquezbernal commented on pkg/tableformatter/terminal.go:
```diff
@@ -0,0 +1,74 @@
+		return
+	}
+
+	if !term.IsTerminal(0) {
```
Maybe use `os.Stdout.Fd()` instead of 0?

mauriciovasquezbernal reviewed (commented):
I like it a lot, the design is clean and well document, great job! It has the potential to avoid a lot of duplicated code. 

I reviewed most of it, looks great. I only have few concerns about the filter, sort and group features. I think those are important, but I'm not sure they're related to a "table writer", for instance, sorting is important in top gadgets even if the output is printed like json. Do you think there is a way to move this logic to a place that is not dependent on the table formatter?

I tried to use this on the top file gadget and it was super simple -> https://github.com/kinvolk/inspektor-gadget/commits/mauricio/tableformatter-top-file. I only hit a problem with the `FileType` member, it's supposed to be a single character but using `width:1` causes it not to be printed. 

Finally, do you think it's possible / worth it to implement a version similar to "tabwriter" to use with gadgets that know all the entries before printing and hence that could adapt the output to better fit the screen? 

btw, the example fails with a wide terminal:

```
tput cols
182
$ go run example.go 
NODE                                           |   PID|COMM  |NAME                                           |                                                                   TIME
-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
Node1                                          |     1|AAA   |Yay                                            |                                                               14677772
panic: runtime error: slice bounds out of range [:69] with length 68

goroutine 1 [running]:
github.com/kinvolk/inspektor-gadget/pkg/tableformatter.(*TableFormatter[...]).BuildFixedString(0xc000124000?, {0x4af6c6, 0x494960?}, 0x47?, 0xc00010cd80, 0x1?)
	/home/mvb/kinvolk/ebpf/inspektor-gadget/pkg/tableformatter/ellipsis.go:38 +0x19a
github.com/kinvolk/inspektor-gadget/pkg/tableformatter.(*TableFormatter[...]).getDefaultFormatter.func1()
	/home/mvb/kinvolk/ebpf/inspektor-gadget/pkg/tableformatter/formatters.go:31 +0x13d
github.com/kinvolk/inspektor-gadget/pkg/tableformatter.(*TableFormatter[...]).FormatEntry(0xc000124000?, 0xc00010ced8)
	/home/mvb/kinvolk/ebpf/inspektor-gadget/pkg/tableformatter/output.go:37 +0x32e
main.main()
	/home/mvb/kinvolk/ebpf/inspektor-gadget/pkg/tableformatter/example/example.go:103 +0x3fd
exit status 2

```

flyth commented on pkg/tableformatter/options.go:
```diff
@@ -0,0 +1,59 @@
+
+type TableFormatterOptions struct {
+	ColumnDivider     string
+	ShowColumnDivider bool
```
Removed it, makes total sense just to use the ColumnDivider.

flyth commented on pkg/tableformatter/columninfo.go:
```diff
@@ -0,0 +1,94 @@
+				ci.EllipsisType = EllipsisStart
+			case "middle":
+				ci.EllipsisType = EllipsisMiddle
+			}
```
I adjusted it to be stricter :)

flyth commented on pkg/tableformatter/columninfo.go:
```diff
@@ -0,0 +1,94 @@
+			}
+			ci.Width = w
+		}
+	}
```
It does now! 👍

flyth commented on pkg/tableformatter/columninfo.go:
```diff
@@ -0,0 +1,94 @@
+	columnType      reflect.Type
+}
+
+func (ci *ColumnInfo[T]) fromTag(tag string) error {
```
I've limited to key and value now - if something is added or wrong, it should be reported now 👍

flyth commented on pkg/tableformatter/columninfo.go:
```diff
@@ -0,0 +1,94 @@
+	tagInfo = tagInfo[1:]
+	for _, subTag := range tagInfo {
+		params := strings.Split(subTag, ":")
+		if len(params) == 1 {
+			switch params[0] {
+			case "ellipsis":
+				ci.EllipsisType = EllipsisEnd
+			case "hide":
+				ci.Visible = false
+			case "fixed":
+				ci.FixedWidth = true
+			}
+			continue
+		}
```
Yeah, makes sense. Refactored 👍

flyth commented on pkg/tableformatter/tableformatter.go:
```diff
@@ -0,0 +1,298 @@
+	return tf, nil
+}
+
+func MustNewTableFormatter[T any](options *TableFormatterOptions) *TableFormatter[T] {
+	tf, err := NewTableFormatter[T](options)
+	if err != nil {
+		panic(err)
+	}
+	return tf
+}
```
I renamed it to MustCreateTableFormatter() for now. I like to use this pattern when initializing things globally. The regex package uses it as well, for example. The advantage of using it that way is that the program crashes instantly and makes you aware of any syntactical mistakes immediately instead of failing (maybe even silently) only when certain code paths are run.

flyth commented on pkg/tableformatter/tableformatter.go:
```diff
@@ -0,0 +1,298 @@
+		tf.getDefaultFormatter(reflect.String, &column)
+	}
+
+	if column.Visible {
+		tf.showColumns = append(tf.showColumns, &column)
+	}
```
Was a double, removed 👍

flyth commented on pkg/tableformatter/tableformatter.go:
```diff
@@ -0,0 +1,298 @@
+		return fmt.Errorf("no extractor set for column %q", column.Name)
+	}
+
+	newColumn := &column
```
Just unclean, fixed 👍

flyth commented on pkg/tableformatter/tableformatter.go:
```diff
@@ -0,0 +1,298 @@
+}
+
+// SetShowColumns takes a comma separated list of column names that will be displayed when using the output methods
+func (tf *TableFormatter[T]) SetShowColumns(columns string) {
```
I changed it to accept []string now. Was thinking back and forth over it, but you're right, it feels cleaner this way when used as a library. 👍

flyth commented on pkg/tableformatter/formatters.go:
```diff
@@ -0,0 +1,50 @@
+		column.Formatter = func(v interface{}) string {
+			return tf.BuildFixedString(v.(string), column.calculatedWidth, column.EllipsisType, column.Alignment)
+		}
+	default:
```
I just missed them. Added 👍 Still unsure about booleans, though.

mauriciovasquezbernal replied:
Why are you unsure about booleans?

flyth replied:
Whether we need them handled separately - right now they are rendered through "%v", so it's "true" and "false" for now. I think I'll add extra support for them later on.

flyth commented on pkg/tableformatter/ellipsis.go:
```diff
@@ -0,0 +1,71 @@
+		return rs
+	}
+
+	if maxLength <= len(EllipsisRune) {
+		return rs[:maxLength]
+	}
+
```
To make sure that when maxLength is 1, we're not just printing the ellipsis, but one character of content instead. I don't know what's better: no content at all or a missing indicator for that there's more content...

mauriciovasquezbernal replied:
I think it's better to print the ellipsis is maxLength is one and there is more than 1 character of content.

flyth commented:
Thanks for the thorough and quick review! That motivates a lot! 👍 
I've addresses most of your feedback already and cleaned up accordingly.

Regarding the additional features (sorting, filtering etc): I agree. This library has become much larger that I anticipated at first and maybe should be renamed to something more generic, but I haven't had a good idea for that, yet ("datautils" might be too generic). I'm sure we could move out certain features to adjacent packages, but I wouldn't remove the dependency from the main package completely, as this would then duplicate code again and decrease convenience again.

I've added `AdjustWidthsToContent()` to do things similar to tabwriter.

The problem with the wide terminal should also be fixed (there was more than one issue open with the auto-scaler).

Again, thanks a lot!

mauriciovasquezbernal commented:
> I'm sure we could move out certain features to adjacent packages, but I wouldn't remove the dependency from the main package completely, as this would then duplicate code again and decrease convenience again.

I agree that we should avoid code duplication. I think it's ok to have "TableFormatter" as the main package and move other (more specific) function to a different package (and maybe in a different PR to make it easier to review?)

flyth commented on cmd/kubectl-gadget/utils/flags.go:
```diff
@@ -133,6 +137,12 @@ func AddCommonFlags(command *cobra.Command, params *CommonFlags) {
 			}
 		}
 
+		// Columns
+		if columns, err := command.PersistentFlags().GetString("columns"); err == nil && columns != "" {
+			// Fallback
+			command.PersistentFlags().Set("output", commonutils.OutputModeCustomColumns+"="+columns)
+		}
```
Yeah, the integration demo is not more than that, a demo. We definitely have to talk about how to integrate it properly with the least code possible. I've also added a description field to the column info, so it would be great to also have an overview of the columns together with a short explanation on `--help`. That and all the new features like sorting, filtering etc (which also partly have to be added to the gadget side of things, then). I'll gladly do a more thorough integration example once the API of the package is stable enough.

blanquicet commented on pkg/columns/formatter/textcolumns/bridge.go:
```diff
@@ -0,0 +1,27 @@
+
+package textcolumns
+
+// This satisfies our current implementation
+
+// TransformEvent is deprecated
+func (tf *TextColumnsFormatter[T]) TransformEvent(event *T) string {
+	return tf.FormatEntry(event)
+}
+
+// BuildColumnsHeader is deprecated
+func (tf *TextColumnsFormatter[T]) BuildColumnsHeader() string {
+	return tf.FormatHeader()
+}
```
This is exactly what will replace the code of the [BaseParser](https://github.com/kinvolk/inspektor-gadget/blob/5c9e9d4db3ee7f98d9a078139c08b8bb54197797/cmd/common/utils/parser.go#L32). It is in charge of calling the `columns` package to achieve what the CLI needs while hiding all the complexity.

blanquicet commented on pkg/columns/sort/sort.go:
```diff
@@ -0,0 +1,144 @@
+		return
+	}
+
+	for i := len(sortBy) - 1; i >= 0; i-- {
```
Put in the function description the order `sortBy` are used.

blanquicet commented on pkg/columns/sort/sort_test.go:
```diff
@@ -0,0 +1,130 @@
+	}
+
+	rand.Shuffle(len(testEntries), func(i, j int) { testEntries[i], testEntries[j] = testEntries[j], testEntries[i] })
+	SortEntries(cols, testEntries, []string{"uint"})
```
Missing test where `sortBy` has more than 1 element.

blanquicet commented on pkg/columns/sort/sort_test.go:
```diff
@@ -0,0 +1,130 @@
+		Float64 float64 `column:"float64"`
+		Bool    bool    `column:"bool"`
+	}
+	testEntries := []*testData{
```
Is it possible to sort by a field that is in an embedded struct? Same test you did for `TestGroupSum`.

flyth replied:
Yes, it's possible. I added a test for that as well 👍

blanquicet commented on pkg/columns/formatter/textcolumns/textcolumns.go:
```diff
@@ -0,0 +1,112 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package textcolumns
```
It is named `textcolumns`, but its main function is `NewFormatter`? I would suggest `formatter` as the package name, don't you think?

flyth replied:
There's also a csv folder (and later on possibly a json folder) with formatters. So I thought csv.NewFormatter() (actually, right now it's csv.NewCSVFormatter() - should rename this to NewFormatter() as well), json.NewFormatter(), textcolumns.NewFormatter() would make sense.

blanquicet commented on pkg/columns/columninfo.go:
```diff
@@ -0,0 +1,196 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package columns
```
I don't know, the "higher" package shouldn't be also the "higher" level of the API? IMO, even if the users call also `columns` directly, the higher level of the API is `textcolumns`, isn't it? In fact, `textcolumns` uses `columns` and not the other way around. Therefore, `columns` should be a sub-package of `textcolumns`. Don't you think?

flyth replied:
The plan was to handle everything aside from the actual tag parsing and preparation of the `Columns` in different feature packages to make it as independent as possible. So every feature relies on the spec from the `Columns` and uses it, yet is completely independent from all other features. An example where it should be used:
We pre-sort things on the gadget side with the top gadgets because we only return a limited number of rows when there are too many. There you would use the sorter (and maybe the filter), but not the formatter (cause that's only used in the "UI" part).

blanquicet replied:
Makes sense!

blanquicet commented on pkg/columns/README.md:
````diff
@@ -0,0 +1,218 @@
+You initialize the formatter by passing it the type of the struct you want to use, like so:
+
+```
+cf, err := tableformatter.NewTableFormatter[Event](nil)
````
Remember to update the documentation with the separated PRs you will create.

blanquicet reviewed (commented):
Thanks for working on this. The result is fantastic!

In this review, I focused on understanding overall how the library works, how it is structured and how it can be integrated with the current code. I will review the code in more detail once the separated implementations are opened.

> This library has become much larger that I anticipated at first and maybe should be renamed to something more generic, but I haven't had a good idea for that, yet ("datautils" might be too generic). 
>
As suggested in the comments, I would call the "entry-point" of the library `formatter` with the following sub-libraries: `columns`, `ellipsis`, `filter`, `group` and `sort`. WDYT?

flyth commented on pkg/columns/ellipsis/ellipsis.go:
```diff
@@ -0,0 +1,60 @@
+		return rs
+	}
+
+	if maxLength <= 1 {
```
Nice catch! 👍

flyth commented on pkg/columns/columns.go:
```diff
@@ -0,0 +1,206 @@
+	return cols
+}
+
+// NewColumns creates a new column helper. T must be of type struct and
+// its fields must contain the column tags.
+func NewColumns[T any](options *ColumnOptions) (*Columns[T], error) {
+	if options == nil {
+		options = GetDefault()
+	}
```
Agree, looks cleaner. I will change it. 👍

flyth commented on pkg/columns/columninfo.go:
```diff
@@ -0,0 +1,196 @@
+				return fmt.Errorf("invalid width %q for field %q: %w", params[1], ci.Name, err)
+			}
+			ci.Width = w
+		default:
```
Great idea to handle it that way. I would probably change or at least default to WithoutKeywords() (maybe name it WithoutTags()), so everything is used by default. This should disable the given columns completely, then, so that even calls to GetColumns() wouldn't return them and you can't use sort etc on it. Maybe we should consider adding columnTags:"abc,def" instead of putting it inside the column:"..." to allow multiple keywords.

Btw: the order defaults to the order of fields in the struct (*10), but an optional "order:number" can be used to reorder it (haven't pushed that change, yet). And then you can always use SetShowColumns() on the formatter to force columns and the order.

blanquicet replied:
> I would probably change or at least default to WithoutKeywords() (maybe name it WithoutTags()), so everything is used by default.
>
I would prefer that the default behaviour is to return the gadget-specific columns. It means, those without keywords and without `hide`. IMO, It would be more intuitive for users.

> This should disable the given columns completely, then, so that even calls to GetColumns() wouldn't return them and you can't use sort etc on it. 
>
I am not sure I follow you. Could you please describe this idea in more detail?

> Maybe we should consider adding columnTags:"abc,def" instead of putting it inside the column:"..." to allow multiple keywords.
>
Yeah, it would be more readable than `column:"name,keyword:kubernetes,keyword:runtime,width:16"`.

---

Regarding the 4th use-case (`local-gadget --k8s-metadata`):

> Btw: the order defaults to the order of fields in the struct (*10), but an optional "order:number" can be used to reorder it (haven't pushed that change, yet).
>
But such order is defined in the struct, so there is no way to have a different order between two callers (`kubectl-gadget` and `local-gadget`).

> And then you can always use SetShowColumns() on the formatter to force columns and the order.
> 
Yeah, I was thinking about something like this:

```golang
func NewTraceParserWithRuntimeInfo(commonFlags *utils.CommonFlags) (TraceParser[types.Event], error) {
	keywords := []string{columns.KeywordContainerRuntime}
	if commonFlags.K8sMetadata {
		keywords = append(keywords, columns.KeywordKubernetes)
	}

	cols, err := columns.NewColumns[types.Event](columns.WithKeywords(keywords...))
	if err != nil {
		return nil, fmt.Errorf("creating columns: %w", err)
	}

	var defColumns []string
	if commonFlags.K8sMetadata {
		// Based on keywords, define the defColumns:
		//  - First the columns with "runtime" keyword: Container runtime, ID, Container
		//  - Then, the columns without keyword (gadget-specific): PID, PPID, PCOMM...
		//  - Finally, the columns with "kubernetes" keyword: Pod, Namespace, Container. BTW, for these, we need to change the name and add the "K8S-" prefix.
		// Example:
		// $ local-gadget trace exec --k8s-metadata
		// CONTAINER RUNTIME    ID        CONTAINER       PID     PPID    PCOMM ... K8S-POD    K8S-NAMESPACE    K8S-CONTAINER
	}

	return textcolumns.NewFormatter(cols, &textcolumns.Options{DefaultColumns: defColumns}), nil
}
```

What do you think?

cc @mauriciovasquezbernal

blanquicet replied:
I noticed something about this proposal. If I am making this call:

```golang
	cols, err := columns.NewColumns[types.Event](columns.WithKeywords(columns.KeywordKubernetes))
```

I would expect to get only the columns with `KeywordKubernetes` keyword but, I was supposing it will return also the columns without tags. Maybe that's why you meant with `WithoutKeyword`. By default, if no options are provided, it will return the column without keywords. Instead, if the user wants columns with Kubernetes and no keyword, they has to do:

```golang
	cols, err := columns.NewColumns[types.Event](columns.WithKeywords(columns.KeywordKubernetes), columns.WithoutKeywords())
```

flyth replied:
I think we got this covered now.

flyth commented:
Thanks for the feedback @blanquicet! 👍 

Regarding the entry point: I've answered in the comment.

mauriciovasquezbernal commented on pkg/columns/filter/filter.go:
```diff
@@ -0,0 +1,273 @@
+				continue
+			}
+
+			// Transform group key according to request
```
Is this comment updated? Seems out of place to me.

mauriciovasquezbernal commented on pkg/columns/filter/filter_test.go:
```diff
@@ -0,0 +1,285 @@
+		{filterString: "string:~(?i)??//{demo", expectedCount: 0, expectError: true}, // garbage regular expression search
+
+		{filterString: "int:", expectedCount: 0, expectError: true},      // match on int, empty string
+		{filterString: "int:1", expectedCount: 1, expectError: false},    // match on int, empty string
```
```suggestion
		{filterString: "int:1", expectedCount: 1, expectError: false},    // match on int, exact match 
```

?

mauriciovasquezbernal commented on pkg/columns/filter/filter_test.go:
```diff
@@ -0,0 +1,285 @@
+		{filterString: "int:>1", expectedCount: 4, expectError: false},   // match on int, gt
+		{filterString: "int:>=1", expectedCount: 5, expectError: false},  // match on int, gte
+		{filterString: "uint:", expectedCount: 0, expectError: true},     // match on uint, empty string
+		{filterString: "uint:1", expectedCount: 1, expectError: false},   // match on uint, empty string
```
ditto.

mauriciovasquezbernal commented on pkg/columns/filter/filter_test.go:
```diff
@@ -0,0 +1,285 @@
+		{filterString: "uint32:~1", expectedCount: 0, expectError: true},   // match on uint32, wrong comparison type
+		{filterString: "uint32:!1", expectedCount: 4, expectError: false},  // match on uint32, negated
+		{filterString: "uint32:<2", expectedCount: 1, expectError: false},  // match on uint32, lt
+		{filterString: "uint32:<=1", expectedCount: 1, expectError: false}, // match on uint32, lte
+		{filterString: "uint32:>1", expectedCount: 4, expectError: false},  // match on uint32, gt
+		{filterString: "uint32:>=1", expectedCount: 5, expectError: false}, // match on uint32, gte
+
+		{filterString: "int64:", expectedCount: 0, expectError: true},      // match on int64, empty string
+		{filterString: "int64:1", expectedCount: 1, expectError: false},    // match on int64, empty string
+		{filterString: "int64:~1", expectedCount: 0, expectError: true},    // match on int64, wrong comparison type
+		{filterString: "int64:!1", expectedCount: 4, expectError: false},   // match on int64, negated
+		{filterString: "int64:<2", expectedCount: 1, expectError: false},   // match on int64, lt
+		{filterString: "int64:<=1", expectedCount: 1, expectError: false},  // match on int64, lte
+		{filterString: "int64:>1", expectedCount: 4, expectError: false},   // match on int64, gt
+		{filterString: "int64:>=1", expectedCount: 5, expectError: false},  // match on int64, gte
+		{filterString: "uint64:", expectedCount: 0, expectError: true},     // match on uint64, empty string
+		{filterString: "uint64:1", expectedCount: 1, expectError: false},   // match on uint64, empty string
+		{filterString: "uint64:~1", expectedCount: 0, expectError: true},   // match on uint64, wrong comparison type
+		{filterString: "uint64:!1", expectedCount: 4, expectError: false},  // match on uint64, negated
+		{filterString: "uint64:<2", expectedCount: 1, expectError: false},  // match on uint64, lt
+		{filterString: "uint64:<=1", expectedCount: 1, expectError: false}, // match on uint64, lte
+		{filterString: "uint64:>1", expectedCount: 4, expectError: false},  // match on uint64, gt
+		{filterString: "uint64:>=1", expectedCount: 5, expectError: false}, // match on uint64, gte
+
+		{filterString: "float32:", expectedCount: 0, expectError: true},     // match on float32, empty string
+		{filterString: "float32:1", expectedCount: 1, expectError: false},   // match on float32, empty string
+		{filterString: "float32:~1", expectedCount: 0, expectError: true},   // match on float32, wrong comparison type
+		{filterString: "float32:!1", expectedCount: 4, expectError: false},  // match on float32, negated
+		{filterString: "float32:<2", expectedCount: 1, expectError: false},  // match on float32, lt
+		{filterString: "float32:<=1", expectedCount: 1, expectError: false}, // match on float32, lte
+		{filterString: "float32:>1", expectedCount: 4, expectError: false},  // match on float32, gt
+		{filterString: "float32:>=1", expectedCount: 5, expectError: false}, // match on float32, gte
+		{filterString: "float64:", expectedCount: 0, expectError: true},     // match on float64, empty string
+		{filterString: "float64:1", expectedCount: 1, expectError: false},   // match on float64, empty string
+		{filterString: "float64:~1", expectedCount: 0, expectError: true},   // match on float64, wrong comparison type
+		{filterString: "float64:!1", expectedCount: 4, expectError: false},  // match on float64, negated
+		{filterString: "float64:<2", expectedCount: 1, expectError: false},  // match on float64, lt
+		{filterString: "float64:<=1", expectedCount: 1, expectError: false}, // match on float64, lte
+		{filterString: "float64:>1", expectedCount: 4, expectError: false},  // match on float64, gt
+		{filterString: "float64:>=1", expectedCount: 5, expectError: false}, // match on float64, gte
```
All these comments could become the test name to pass to `t.Run()` below. (same comment applies to other parts on the code).

flyth replied:
Moved from comment to field for a couple of tests now => tests much more readable.

mauriciovasquezbernal reviewed (approved):
It's a great and huge PR. I tried to review as much as possible but it's very difficult to review such a small PR taking into consideration all the details. I think the code is in a very good shape and we can merge it.

flyth commented on pkg/columns/columninfo_test.go:
```diff
@@ -0,0 +1,444 @@
+		Field string `column:"fail,align:left:bar"`
+	}
+
+	_, err := NewColumns[testSuccess1]()
+	if err != nil {
+		t.Errorf("failed to initialize: %v", err)
+	}
```
Yeah, I had those tests prior to the refactoring, will re-add them. 👍

flyth commented on pkg/columns/columns_test.go:
```diff
@@ -0,0 +1,317 @@
+	if ocols[1].Name != "stringField" {
+		t.Errorf("expected second entry to be stringField")
+	}
+	if len(ocols) != 2 {
+		t.Errorf("expected two columns")
+	}
```
I will move it (makes more sense), but I'd need to return as well to avoid the panic.

mauriciovasquezbernal replied:
use `t.Fatal()`?

flyth replied:
I changed it in the last commit already.

alban commented on pkg/columns/README.md:
```diff
@@ -0,0 +1,102 @@
+# Columns Library
+
```
This package documentation could be in `pkg/columns/doc.go` and `pkg/columns/example_test.go`, so that go doc can pick it up.

Example:
* Package documentation: https://pkg.go.dev/encoding/gob
* Source `doc.go`: https://go.dev/src/encoding/gob/doc.go?m=text
* Source `example_{,*_}test.go`: https://go.dev/src/encoding/gob/example_test.go?m=text

flyth replied:
Good point! I moved most of the documentation and the most important examples now. 👍

blanquicet commented on pkg/columns/formatter/textcolumns/options.go:
```diff
@@ -0,0 +1,93 @@
+		ColumnDivider:  DividerSpace,
+		DefaultColumns: nil,
+		HeaderStyle:    HeaderStyleUppercase,
+		RowDivider:     DividerDash,
+		ShowRowDivider: true,
```
I `ShowRowDivider` really needed? Isn't it enough to set `RowDivider` to `NoDivider` (To be added)?

flyth replied:
Removed it! 👍

blanquicet commented on pkg/columns/formatter/textcolumns/options.go:
```diff
@@ -0,0 +1,93 @@
+
+func DefaultOptions() *Options {
+	return &Options{
+		AutoScale:      true,
+		ColumnDivider:  DividerSpace,
+		DefaultColumns: nil,
+		HeaderStyle:    HeaderStyleUppercase,
+		RowDivider:     DividerDash,
+		ShowRowDivider: true,
```
I think the PR is ready to be merged, my main concern are the defaults values here:

- Do we want to start using a divider by default? 
- Do we want to autoscale by default?

I don't know, I think this PR should keep the same defaults for the UI and, in case, we can change those in the future actually to take advantage of all the features this library offers. 

cc @mauriciovasquezbernal @alban

flyth replied:
Agree, we should use what we have now and maybe change it later on. I set the defaults to AutoScale = false and RowDivider = DividerNone now. 👍

mauriciovasquezbernal replied:
I agree. Let's keep the same output we have now and then we can switch to something else.

blanquicet commented on pkg/columns/options.go:
```diff
@@ -0,0 +1,63 @@
+	DefaultAlignment Alignment             // default text alignment to use
+	DefaultEllipsis  ellipsis.EllipsisType // default type of ellipsis to use for overflowing text
+	DefaultWidth     int                   // width to be used when no width is specified for a column
+	RequireColumnTag bool                  // if set to true, Columns will consider all struct members, regardless of the column tag being present
```
```suggestion
	RequireColumnTag bool                  // if set to false, Columns will consider all struct members, regardless of the column tag being present
```
?

blanquicet commented on pkg/columns/options.go:
```diff
@@ -0,0 +1,63 @@
+func GetDefault() *Options {
+	return &Options{
+		DefaultAlignment: AlignLeft,
+		DefaultEllipsis:  ellipsis.End,
```
Do we want to add the ellipsis at the end by default?

cc @mauriciovasquezbernal @alban

flyth replied:
So, are we fine with the ellipsis at the end of the string by default? I think it's more useful to know that content has been truncated and losing one character than not knowing.

mauriciovasquezbernal replied:
I think ellipsis at the end is ok by default. But, wouldn't it affect the integration tests?

flyth replied:
Yeah, but truncating without the ellipsis would probably affect them as well.

blanquicet replied:
Yes, it does. It is the reason of https://github.com/kinvolk/inspektor-gadget/pull/934 CI failing:

```bash
NODE             NAMESPACE        POD                            CONTAINER                 PID     TPID    SIGNAL           RET    COMM            
fv-az28-19       test-sigsnoop-9… test-pod                       test-pod                  15269   15269   SIGURG           0      runc:[2:INIT]   
fv-az28-19       test-sigsnoop-9… test-pod                       test-pod                  15269   1       SIGURG           0      runc:[2:INIT]   
fv-az28-19       test-sigsnoop-9… test-pod                       test-pod                  15269   15283   SIGTERM          0      sh              
```

blanquicet replied:
I tried, and it is truncated. Therefore, I agree `ellipsis.End` is the correct way to go. 

At this point, given that the library is truncating by default, it should indeed autoscale by default too. By doing so, the columns will be truncated only if the terminal is not big enough.

Sorry @flyth for asking you to disable and now enable the autoscaling. I was missing the fact that the library is truncating by default 🙈.

blanquicet commented on pkg/columns/examples/gadget/gadget.go:
```diff
@@ -0,0 +1,107 @@
+	fmt.Println()
+
+	// Leave out kubernetes info for this one, but include gadget data (not-embedded struct) and runtime information
+	formatter = textcolumns.NewFormatter(
+		gadgetColumns.GetColumnMap(columns.Or(columns.WithEmbedded(false), columns.WithTag("runtime"))),
+		textcolumns.WithAutoScale(false),
+	)
```
I think we still need a way to gather only the gadget-specific columns. We cannot rely on the fact that the gadget-specific fields are not embedded. Also because the some gadgets have embedded fields, e.g. [top ebpf gadget](https://github.com/kinvolk/inspektor-gadget/blob/ad87c3f5407ae5bd30d9fc486e79df4662ec19ae/pkg/gadgets/top/ebpf/types/types.go#L119).

Maybe implementing a `WithNoTags()` filter? Anyway, this is more about the usage and not the actual implementation of the library so we can go ahead with this PR and manage this concern in a second PR.

flyth replied:
WithNoTags() is a great idea to do that, I added the function from your PR as discussed, so it's now part of the library. 👍

blanquicet reviewed (commented):
I have just some concerns about the default values, but in general, I really like it! 

BTW, I started working on the integration, and it is removing all the repeated/annoying code! 🎉 🎉

flyth commented on pkg/columns/README.md:
```diff
@@ -0,0 +1,102 @@
+
+### Options
+
+| Field            | Value | Description                                                         |
+|------------------|-------|---------------------------------------------------------------------|
+| DefaultAlignment |       | default alignment for a column to use if not specified for a column |
+| DefaultWidth     | int   | width to be used when no width is specified for a column            |
+| DefaultWidth     | int   | width to be used when no width is specified for a column            |
```
Added default values as comments - now handled by go doc.

flyth commented on pkg/columns/options.go:
```diff
@@ -0,0 +1,63 @@
+		DefaultAlignment: AlignLeft,
+		DefaultEllipsis:  ellipsis.End,
+		DefaultWidth:     16,
+		RequireColumnTag: true,
```
Yeah, it's confusing - do you think it would lessen the confusion, if we would rather name that parameter "RequireColumnDefinition"?

blanquicet replied:
Yeah, it could. But the most important thing is to add a comment on the parameter definition explaining that it refers to the `columns` JSON tag.

blanquicet commented on pkg/columns/columns.go:
```diff
@@ -0,0 +1,262 @@
+func MustCreateColumns[T any](options ...Option) *Columns[T] {
+	cols, err := NewColumns[T](options...)
+	if err != nil {
+		panic(err)
```
Is it possible to specify in the error message the type for which this function failed?

blanquicet commented on pkg/columns/columns.go:
```diff
@@ -132,6 +132,24 @@ func (c ColumnMap[T]) GetColumnNames(filters ...ColumnFilter) []string {
+// VerifyColumNames takes a list of column names and returns two lists, one containing the valid column names
+// and another containing the invalid column names. ignorePrefixes defines whether prefixes like "-" for descending
+// sorting should be ignored.
+func (c ColumnMap[T]) VerifyColumNames(columnNames []string, ignorePrefixes bool) (valid []string, invalid []string) {
```
```suggestion
func (c ColumnMap[T]) VerifyColumnNames(columnNames []string, ignorePrefixes bool) (valid []string, invalid []string) {
```

blanquicet reviewed (approved):
Thanks @flyth for managing all the comments! I think this library will introduce a huge improvement to the CLI, and 3rd-party app using the tracers and trying to format it!

LGTM! 🎉 🎉

flyth commented on pkg/columns/columninfo.go:
```diff
@@ -0,0 +1,221 @@
+			if err != nil {
+				return fmt.Errorf("invalid precision value %q for field %q: %w", params[1], ci.Name, err)
+			}
+			if w < 0 {
```
Yeah, though actually it's a typo :) I planned to use `< -1`, since strconv.FormatFloat() - which we are using - allows the special precision of `-1` to format a float as precise as possible.

flyth commented on pkg/columns/columns.go:
```diff
@@ -0,0 +1,246 @@
+		return errors.New("no name set for column")
+	}
+
+	columnName := strings.ToLower(column.Name)
```
Yeah, we could, but it wouldn't safe us from calling strings.ToLower() anyways, so this way it at least doesn't happen twice.

flyth commented on pkg/columns/ellipsis/README.md:
```diff
@@ -0,0 +1,14 @@
+| Foobar123    | 4            | None           | Foob   |
+| Foobar123    | 4            | End            | Foo…   |
+| Foobar123    | 4            | Start          | …123   |
+| Foobar123    | 4            | Middle         | Fo…r   |
```
Yeah, good catch! I removed that though with the markdown docs and changed it to using an example for go doc defined here: https://github.com/kinvolk/inspektor-gadget/pull/877/files#diff-0eea058f088deef22cbf26ff7010c183d813761e0400cb542f3acff26961b53dR23

flyth commented:
Thanks everyone for reviewing this big one! As said, it has become much larger than anticipated and next time I'll definitely split it up more (and just for everyone to be on the same page: the last decision we agreed upon was not to split it up anymore). I know PRs this huge are tiresome, so I really appreciate the effort and hope this lib will save the time afterwards :)

@eiffel-fl I will squash the commits so you get a better view.

flyth commented on pkg/columns/README.md:
```diff
@@ -0,0 +1,102 @@
+|------------------|-------|---------------------------------------------------------------------|
+| DefaultAlignment |       | default alignment for a column to use if not specified for a column |
+| DefaultWidth     | int   | width to be used when no width is specified for a column            |
+| DefaultWidth     | int   | width to be used when no width is specified for a column            |
```
This already got removed and is now filled by go doc automatically (and correctly) from the Options struct.

flyth commented on pkg/columns/examples/gadget/gadget.go:
```diff
@@ -0,0 +1,107 @@
+var GadgetOutput = []*GadgetData{
+	{
+		Kubernetes: Kubernetes{
+			Node:      "Node 1",
```
Haven't thought about doing that in tests, good idea - however, as this is just an example, I think it might confuse more than help when creating the entries somewhere hidden in the code at runtime. If you worry about the line count, maybe we should just remove the "Node 3" entry.

flyth commented on pkg/columns/filter.go:
```diff
@@ -0,0 +1,113 @@
+	}
+	numTags := len(tags)
+	return func(matcher ColumnMatcher) bool {
+		marked := make([]bool, numTags)
```
Not sure what I was thinking there - it's really way too complicated... 👍

flyth commented on pkg/columns/filter/examples_test.go:
```diff
@@ -0,0 +1,89 @@
+}
+
+func ExampleGetFilterFromString() {
+	type Employee struct {
+		Name       string `column:"name" columnTags:"sensitive"`
+		Age        int    `column:"age" columnTags:"sensitive"`
+		Department string `column:"department"`
+	}
```
It's what lands in the example code blocks on the go doc page - honestly I haven't tested how well it does if there's globals defined (I think we want the struct definitions to show up in the example code). I will check that.

flyth commented on pkg/columns/filter/examples_test.go:
```diff
@@ -0,0 +1,89 @@
+
+	// Output:
+	// {Alice 32 Security}
+	// {Bob 26 Security}
```
This is only meant to be an example for the go doc page - with the additional benefit that it gets tested alongside all other tests. But for that to succeed or fail, you have to add these `// Output:` lines and it will compare the stdout with those comments (you don't have `*testing.T` in here).

flyth commented on pkg/columns/filter/filter.go:
```diff
@@ -0,0 +1,262 @@
+}
+
+// FilterEntries will filter entries according to the given filters.
+func FilterEntries[T any](cols columns.ColumnMap[T], entries []*T, filters []string) ([]*T, error) {
```
Yes.

flyth commented on pkg/columns/group/group_test.go:
```diff
@@ -0,0 +1,141 @@
+		},
+		{
+			Name:    "GroupByColumn",
+			GroupBy: []string{"name"},
```
I will add a test for that 👍
